keras/backend/mxnet_backend.py

```diff
@@ -2690,55 +2654,394 @@ def rnn(step_function, inputs, initial_states,
+            masks = [None for _ in inputs]
+
+        # Iterate over a time step
+        for inp, msk in zip(inputs, masks):
+            last_output, new_states = step_function(KerasSymbol(inp),
+                                                    states + constants)
+            if getattr(last_output, '_uses_learning_phase', False):
+                uses_learning_phase = True
+            if msk is not None:
+                new_states = [KerasSymbol(mx.sym.where(msk,
+                                                       ns.symbol,
+                                                       s.symbol))
+                              for s, ns in zip(states, new_states)]
+                # Initialize the output for first time step
+                if prev_output is None:
+                    prev_output = zeros_like(last_output)
+                last_output = KerasSymbol(mx.sym.where(msk,
+                                                       last_output.symbol,
+                                                       prev_output.symbol))
+                prev_output = last_output
+            states = new_states
+            # Expand the output dimension from `(samples, output_dim)` to
+            # `(samples, 1, output_dim)` with middle axis as time dimension
+            outputs.append(mx.sym.expand_dims(last_output.symbol, axis=1))
+        # Concatenate the output across time dimension
+        outputs = mx.sym.concat(*outputs, dim=1)
     else:
-        masks = [None for _ in inputs]
+        # TODO: remove version check after mxnet 1.3.1 release
+        if mx.__version__ != '1.3.1':
+            raise NotImplementedError('unroll=False in RNN only works with MXNet 1.3.1 or newer ,'
+                                      ' please upgrade to latest master using: pip install --ugprade mxnet --pre')
+        # Reverse the input sequence
+        if go_backwards:
+            inputs = reverse(inputs, 0)
+            mask = reverse(mask, 0)
+
+        # Transpose to time-major, i.e.
+        # from (batch, time, ...) to (time, batch, ...)
+        ndim = len(inputs.shape)
+        axes = [1, 0] + list(range(2, ndim))
+        inputs = mx.sym.transpose(inputs.symbol, axes=axes)
+
+        if mask is not None:
+            if len(mask.shape) == len(dshape) - 1:
+                mask = expand_dims(mask)
+            mask = mx.sym.transpose(mask.symbol, axes=axes)
+            data = [inputs, mask]
+        else:
+            data = [inputs]
+        states = [state.symbol for state in initial_states] + [constant.symbol for constant in constants]
+
+        # using control flow operators(foreach) in MXNet if not unrolling
+        # foreach operator only take step functions with MXNet symbol as input
+        # translating step function from Keras RNN cells to pure MXNet
+        # define _step according to different type of RNN Cells
+        if type(cell) == keras.layers.recurrent.SimpleRNNCell:
+            def _step(data, states):
```

Comment:
Please move these to separate functions for easy maintenance.

def _step_for_simple_rnn(), _step_for_lstm() .. etc.